### DISCUSSION TITLE
TypeConverter and clarification on @embedded

### DISCUSSION THREAD

**Discussion by Mattertorian** [isar/isar#730] [Ideas]
First of all, thank you for this great database!

In v3, TypeConverters were removed in favor of `@embedded` and `@enumerated`. These new options are great but I do not think they sufficiently replace the TypeConverter (or perhaps I do not fully understand their use).

Suppose a class like the following:

`class Person {
  final String firstName;
  final Duration attentionSpan;
}`

With TypeConverter, this could be handled with a simple `@DurationConverter()` annotation. To my understanding, existing classes like `Duration` cannot be `@embedded`
In another thread, it was mentioned that the correct approach would be to add something like
`int get isarDuration => attentionSpan.microseconds`
and 
`set isarDuration(int microseconds) => attentionSpan = Duration(microseconds: microseconds)`

However, this approach does not allow for immutable models as the `attentionSpan` property can no longer be marked as `final`.

Am I misunderstanding the use of `@embedded`? And if not, is it possible to bring back TypeConverters?

**Comment by Diamondemon** [comment]
I have the same question.
I would like to be able to embed Duration objects but I cannot understand how. Is there a new way to do it?

**Comment by renanzdm** [comment]
You don't need to save a Duration class which is a Dart specific object, you can simply save an integer and do conversions to use in your logic. This loosens the coupling of your data classes

**Reply by maeddin** [reply]
This way you would have to adjust the logic every time you change which data type is stored in the background. With TypeConverters you could sort all this out completely independently of the logic. So in my opinion this approach creates an unnecessary dependency on the logic and therefore doesn't really support decoupling.

**Comment by maeddin** [comment]
I delayed upgrading to Isar 3 until now because I had to rely on the `TypeConverters`. With Dart 3, I'm now forced to upgrade Isar as well, which means I have to throw a lot of models out the window. For some models, for example, I'm not willing to have no `required` parameters because for some I want to force something to be specified.
And Dart objects like `Duration` and `Color` can't be saved now either, which makes no sense to me. This converting between Dart and Isar types was well encapsulated with `TypeConverters`.
But since I think there is some reason behind this, I would be interested in a more detailed explanation of this big breaking change.

**Reply by Sharkgamedev** [reply]
I am running into this problem, too, with Timestamps. My solution for the time being is much like what others in this thread have suggested, to store an isar-compatible format but keep the original field (with @ignore) using a getter to 'convert' on the fly... Not a huge fan of this as TypeConverters seem like they would have been a great option here?

Am I the stupid one here for trying to store a complex object from a package I have no control over in Isar? I assume that isar's design considers this a non-problem so I too am curious to hear the thought process behind that.

That being said, it slows me down a little sure but it is a non-critical issue that may have been addressed in v4.0 and I am grateful for the work put into v3 for free ♥ certainly I could fork the repo if I really was all that torn up over it :)

**Comment by Ilusinusmate** [comment]
Currently running into the same problem, but mine is not something like a duration that you can easily  build a wrapper for an integer and get it to work, I need to deal with Decimals for money precision, my app needs to run both online and offline, so storing this Decimals into local device storage it's crucial for the whole app worflow. Please create some support for strutured Classes. I've tried to create setters and getters to wrap a String, but Isar doesnt deal well with it when trying to update the record, it just doesnt update or create it, it just sets the default value I've set on the collection schema.
